### PR TITLE
vLLM v1

### PR DESCRIPTION
**Pull request by yahya010**
Update to vLLM v1

### CODE REVIEW

**Review by shepardxia — COMMENTED** [review]
Format is a bit weird, are you using pre-commit hooks? Might be worth checking the codecov too. But otherwise backend itself looks good to me!

File: genlm/backend/llm/vllm.py [hunk 1]
```diff
@@ -108,51 +169,55 @@ def from_name(cls, model_name, engine_opts=None, **kwargs):
...
-                        "Setting enable_chunked_prefill to True may interfere with AsyncVirtualLM's "
-                        "custom sampling functionality."
-                    )
-
             engine_opts = {
                 "enable_prefix_caching": True,
-                "disable_log_requests": True,
-                "disable_async_output_proc": True,  # This parameter forces vLLM to use v0, which is currently what we want to do.
+                "disable_log_stats": True,
+                "gpu_memory_utilization": 0.5,
```
Comment (vicky-xef) [inline genlm/backend/llm/vllm.py]: why is this set to 0.5?

File: genlm/backend/llm/vllm.py [hunk 2]
```diff
@@ -40,22 +46,64 @@ def from_name(cls, *args, **kwargs):  # pragma: no cover
...
+                if self._captured_batch is None:
+                    return None
+                return self._captured_batch.clone()
+
+        def clear(self):
+            """Clear captured logprobs."""
+            with self._lock:
+                self._captured_batch = None
+
+    class AsyncVirtualLM(AsyncLM):  # pragma: no cover
```
Comment (vicky-xef) [inline genlm/backend/llm/vllm.py:139]: no coverage here?

File: genlm/backend/llm/vllm.py [hunk 3]
```diff
@@ -40,22 +46,64 @@ def from_name(cls, *args, **kwargs):  # pragma: no cover
...
+            self.device = device
+            self._captured_batch = None  # [batch_size, vocab_size] tensor
+            self._lock = threading.Lock()
+
+        def apply(self, logits: torch.Tensor) -> torch.Tensor:
+            """Capture logprobs and pass through logits unchanged."""
+            logprobs = torch.log_softmax(logits, dim=-1, dtype=logits.dtype)
+            with self._lock:
+                # Single clone of entire batch - O(1) instead of O(batch_size)
+                self._captured_batch = logprobs.clone()
```
Comment (samuki) [inline genlm/backend/llm/vllm.py]: Could it ever happen that this step overwrites logprobs if `apply()` is called multiple times per step? Or is it guaranteed that we call it only once?

File: benchmark/benchmark_v0_v1.py [hunk 4]
```diff
@@ -0,0 +1,370 @@
+#!/usr/bin/env python3
```
Comment (samuki) [inline benchmark/benchmark_v0_v1.py:1]: @yahya010 do you have some numbers from the benchmark?

File: genlm/backend/llm/vllm.py [hunk 5]
```diff
@@ -250,49 +405,89 @@ def batch_next_token_logprobs_sync(self, token_ids_list):
...
+
+                # Clear CUDA cache
+                if torch.cuda.is_available():
+                    torch.cuda.empty_cache()
+                    torch.cuda.synchronize()
+
+                # Clean up distributed state
                 destroy_model_parallel()
                 destroy_distributed_environment()
+            except Exception:
```
Comment (samuki) [inline genlm/backend/llm/vllm.py]: What exceptions could we get here? It would be better to catch the specific exception.

File: benchmark/benchmark_v0_v1.py [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+            f"  Avg latency: {result.avg_latency_ms:.2f}ms ± {result.std_latency_ms:.2f}ms"
+        )
+        print(f"  Per-request: {result.per_request_ms:.2f}ms")
+        print(f"  Throughput:  {result.throughput_rps:.1f} req/s")
+        print(f"  GPU Memory:  {result.gpu_memory_mb:.1f} MB")
+
+    return results
+
+
+def print_comparison(
```
Comment (samuki) [inline benchmark/benchmark_v0_v1.py:272]: This function is never called

File: genlm/backend/llm/vllm.py [hunk 7]
```diff
@@ -40,22 +46,64 @@ def from_name(cls, *args, **kwargs):  # pragma: no cover
...
-            self.log_probs = None
-
-        def __call__(self, past_token_ids, logits):
-            assert self.log_probs is None, (
-                "Log probs already set. This should never happen."
-            )
-            self.log_probs = torch.log_softmax(logits, dim=-1, dtype=logits.dtype)
+    logging.getLogger("vllm").setLevel(logging.WARNING)
+
+    class GlobalLogprobsCapture(LogitsProcessor):  # pragma: no cover
```
Comment (samuki) [inline genlm/backend/llm/vllm.py:66]: Could we add a test instead of skipping it?

Comment (yahya010) [reply genlm/backend/llm/vllm.py:66]: no, the CI T4 GPU doesn't work for this

File: genlm/backend/llm/vllm.py (hunk 2 above)
Comment (yahya010) [reply genlm/backend/llm/vllm.py:139]: the CI T4 GPU doesn't work for this

File: genlm/backend/llm/vllm.py [hunk 8]
```diff
@@ -1,21 +1,27 @@
...
 
 from genlm.backend.llm.base import AsyncLM
 from genlm.backend.cache import OutputCache
 
 try:
-    from vllm import AsyncLLMEngine, SamplingParams, AsyncEngineArgs
+    # Enable vLLM v1 with in-process mode (no multiprocessing)
+    # Must be set BEFORE importing vllm
+    os.environ.setdefault("VLLM_USE_V1", "1")
+    os.environ.setdefault("VLLM_ENABLE_V1_MULTIPROCESSING", "0")
```
Comment (ClementeP) [inline genlm/backend/llm/vllm.py]: shoukd this be hard-set? if someone already has VLLM loaded in their environment wth multiprocessing=1, this may not override it.

File: genlm/backend/llm/vllm.py (hunk 5 above)
Comment (ClementeP) [reply genlm/backend/llm/vllm.py]: agree

File: genlm/backend/llm/vllm.py [hunk 9]
```diff
@@ -64,20 +112,31 @@ class AsyncVirtualLM(AsyncLM):
...
             """Initialize an `AsyncVirtualLM` instance.
 
             Args:
-                async_llm_engine (AsyncLLMEngine): The async vLLM engine instance.
+                llm_engine (LLM): The vLLM engine instance.
+                logprobs_capture (GlobalLogprobsCapture): The global logprobs capture processor.
                 cache_size (int, optional): Maximum size of the output cache. If 0, caching is disabled. Defaults to 0.
                 cache_opts (dict, optional): Additional options to pass to the [`OutputCache`][genlm.backend.cache.OutputCache] constructor. Defaults to {}.
+                batch_size (int, optional): Maximum queries to process in one batch during auto-batching. Defaults to 20.
+                timeout (float, optional): Seconds to wait since last query before processing current batch. Defaults to 0.02.
```
Comment (ClementeP) [inline genlm/backend/llm/vllm.py]: I am a bit confused by this ... shouldn't we start counting the time from the *first* query received?

File: genlm/backend/llm/vllm.py [hunk 10]
```diff
@@ -171,74 +236,164 @@ def set_lora(self, lora_path, lora_name='lora_1'):
...
+            result = await future
 
             if self.cache is not None:
                 self.cache[key] = result
 
             return result
 
-        async def _next_token_logprobs(self, token_ids):
-            """Request log probabilities of next token asynchronously.
+        def _add_query(self, token_ids, future):
```
Comment (ClementeP) [inline genlm/backend/llm/vllm.py:317]: Same concern here about what I pinned at line 132. Don't we want to fire after some "x" time since the first query is received?

**Review by ClementeP — COMMENTED** [review]
This is very good! Just left a few comments to address/things to clarify

File: benchmark/benchmark_v0_v1.py (hunk 4 above)
Comment (yahya010) [reply benchmark/benchmark_v0_v1.py:1]: Here are v0 vs v1 numbers on gpt2 / H100, 20 timed iters per batch:

Batch | v0 latency | v1 latency | Speedup | v0 throughput | v1 throughput
-- | -- | -- | -- | -- | --
1 | 8.10 ms | 6.05 ms | 1.34× | 124 req/s | 165 req/s
4 | 32.17 ms | 6.32 ms | 5.09× | 124 req/s | 633 req/s
8 | 65.01 ms | 6.75 ms | 9.64× | 123 req/s | 1186 req/s
16 | 128.22 ms | 7.51 ms | 17.08× | 125 req/s | 2131 req/s
32 | 254.06 ms | 8.97 ms | 28.31× | 126 req/s | 3566 req/s


v0 throughput is flat (~124 req/s) because per-request logits_processors in v0 can't actually share a batch, the wrapper ends up running sequentially. v1's global logits processor + auto-batching scales linearly.

**Review by ClementeP — APPROVED** [review]

File: genlm/backend/llm/vllm.py [hunk 11]
```diff
@@ -64,29 +151,43 @@ class AsyncVirtualLM(AsyncLM):
...
+                llm_engine (LLM): The vLLM engine instance.
+                logprobs_capture (GlobalLogprobsCapture): The global logprobs capture processor.
                 cache_size (int, optional): Maximum size of the output cache. If 0, caching is disabled. Defaults to 0.
-                cache_opts (dict, optional): Additional options to pass to the [`OutputCache`][genlm.backend.cache.OutputCache] constructor. Defaults to {}.
+                cache_opts (dict, optional): Additional options to pass to the [`OutputCache`][genlm.backend.cache.OutputCache] constructor. Defaults to None (no extra options).
+                batch_size (int, optional): Maximum queries to process in one batch during auto-batching. Defaults to 20.
+                timeout (float, optional): Seconds to wait after the first queued query before processing the current batch. The batch also fires immediately when ``batch_size`` is reached. Defaults to 0.02.
 
             Note:
                 The cache stores the log probabilities for previously seen token sequences to avoid redundant requests. KV caching is handled internally by the vLLM engine.
```
Comment (postylem) [inline genlm/backend/llm/vllm.py:174]:
```suggestion
                The cache stores the log probabilities for previously seen token sequences to avoid redundant requests. KV caching is handled internally by the vLLM engine.
                ``batch_next_token_logprobs_sync`` bypasses this cache and always re-evaluates; the other three logprobs methods consult it.
```

File: genlm/backend/llm/vllm.py [hunk 12]
```diff
@@ -250,49 +453,121 @@ def batch_next_token_logprobs_sync(self, token_ids_list):
             Returns:
                 (torch.Tensor): A tensor of normalized log probability tensors, one for each prompt in the input list.
             """
```
Comment (postylem) [inline genlm/backend/llm/vllm.py:461]:
```suggestion
            Returns:
                (torch.Tensor): A tensor of normalized log probability tensors, one for each prompt in the input list.

            Note:
                This method does not consult the output cache (unlike the async batch path,
                which delegates to the cached ``next_token_logprobs``). Every prompt is
                re-evaluated.
            """
```

**Review by postylem — COMMENTED** [review]
Claude caught the following, and it seems worthy of noting in docstrings, if relevant @yahya010 : 

> One spot where cache semantics diverge between the four `*_logprobs*` methods, so a reader doesn't have to infer it from the code.
> `batch_next_token_logprobs_sync` (the override here) bypasses `self.cache`, while the async batch path (`batch_next_token_logprobs`, inherited from `AsyncLM`) transparently uses the cache because it does `asyncio.gather(*[self.next_token_logprobs(t) …])`. The two inline suggestions below capture that asymmetry in the two places a reader is likely to look.
